pkg/s3repository/repository_protected_entity_manager.go

```diff
@@ -16,6 +16,94 @@
 
 package s3repository
 
+import (
+	"context"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/s3"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"github.com/vmware-tanzu/astrolabe/pkg/astrolabe"
+	"strings"
+	"sync"
+)
+
 type RepositoryProtectedEntityManager struct {
-	petms map[string]ProtectedEntityTypeManager
+	petms   map[string]astrolabe.ProtectedEntityTypeManager
+	session session.Session
+	bucket  string
+	prefix  string
+	logger  logrus.FieldLogger
+	mutex   sync.Mutex
+}
+
+func NewS3RepositoryProtectedEntityManager(session session.Session, bucket string,
+	prefix string, logger logrus.FieldLogger) (*RepositoryProtectedEntityManager, error) {
+	s3Client := s3.New(&session)
+	delimiter := "/"
+	if !strings.HasSuffix(prefix, "/") {
+		prefix = prefix + "/"
+	}
+	loi := s3.ListObjectsV2Input{
+		Bucket:    &bucket,
+		Delimiter: &delimiter,
+		Prefix:    &prefix,
+	}
+	loo, err := s3Client.ListObjectsV2(&loi)
+	if err != nil {
+		return nil, errors.WithMessage(err, "Failed retrieving prefixes")
+	}
+	returnPEM := RepositoryProtectedEntityManager{
+		petms:   make(map[string]astrolabe.ProtectedEntityTypeManager),
+		session: session,
+		bucket:  bucket,
+		prefix:  prefix,
+		logger:  logger,
+	}
+	for _, curPrefix := range loo.CommonPrefixes {
+		curPrefixStr := curPrefix.String()
+		components := strings.Split(curPrefixStr, "/")
+		curTypeName := components[len(components)-2]
+		curPETM, err := NewS3RepositoryProtectedEntityTypeManager(curTypeName, session, bucket, prefix, logger)
+		if err != nil {
+
+		} else {
+			returnPEM.petms[curTypeName] = curPETM
+		}
+	}
+
+	return &returnPEM, nil
+}
+
+func (recv RepositoryProtectedEntityManager) GetProtectedEntity(ctx context.Context, id astrolabe.ProtectedEntityID) (astrolabe.ProtectedEntity, error) {
+	petm := recv.GetProtectedEntityTypeManager(id.GetPeType())
+	if petm == nil {
+		return nil, errors.Errorf("No PETM for type %s, cannot retrieve ID %s", id.GetPeType(), id.String())
+	}
+	return petm.GetProtectedEntity(ctx, id)
+}
+
+func (recv RepositoryProtectedEntityManager) GetProtectedEntityTypeManager(peType string) astrolabe.ProtectedEntityTypeManager {
+	recv.mutex.Lock()
+	defer recv.mutex.Unlock()
+	returnPETM := recv.petms[peType]
+	if returnPETM == nil {
+		var err error
```

Comment:
probably don't need to declare the error var

Reply:
It does need to be declared.  It's a go-ism.  When I used returnPETM,err := ... it makes a new variable masking the returnPETM that is declared outside the block.